java/src/main/java/io/cucumber/cucumberexpressions/CucumberExpression.java

```diff
@@ -61,9 +81,18 @@ private String rewriteToRegex(Node node) {
     }
 
     private static String escapeRegex(String text) {
-        return ESCAPE_PATTERN.matcher(text).replaceAll("\\\\$0");
+        int length = text.length();
+        StringBuilder sb = new StringBuilder(length * 2); // prevent resizes
+        int maxChar = CHAR_TO_ESCAPE.length;
+        for (int i = 0; i < length; i++) {
```

Comment:
I think that in practice expressions with characters that need to be escaped are rare. Would it make sense to check the entire string for characters that need to be replaced first? It would save an array copy.

Reply:
If we rarely escape in practice, the situation is a bit different. We need to benchmark the variants based on the escaping frequency, not only on pure escaping performance.

I've implemented two other variants, so to summarize:

| Variant | Description |
| -- | -- |
| escapeRegex0 | Pattern (cucumber 7.11.0) |
| escapeRegex1 | append(char) with 1 array allocation and 1 array copy |
| escapeRegex2 | append(String) with 1 array allocation and 1 array copy |
| escapeRegex3 | append(char) with 1 array allocation and 0..1 array copy |
| escapeRegex4 | append(String) with 0..1 array allocation and 0..1 array copy |

I benchmarked these variants with JMH for varying escaping frequencies:

![Cucumber_escapeRegex_performance](https://user-images.githubusercontent.com/2843894/212463350-79924b97-41b9-47a4-a725-14cf07437b96.png)

The variant `escapeRegex4` outperforms the other variants for all escaping frequencies. Depending on escaping frequency, this variant is 1.7 to 2.9 times faster the cucumber 7.11.0 implementation. The code is a bit more complex, but still understandable, so I think we can use this variant:

`public static String escapeRegex4(String text) {
    int length = text.length();
    StringBuilder sb = null;
    int blocStart=0;
    int maxChar = CHAR_TO_ESCAPE.length;
    for (int i = 0; i < length; i++) {
        char currentChar = text.charAt(i);
        if (currentChar < maxChar && CHAR_TO_ESCAPE[currentChar]) {
            if (sb == null) {
                sb = new StringBuilder(length * 2);
            }
            if (i > blocStart) {
                // flush previous block
                sb.append(text, blocStart, i);
            }
            sb.append('\\');
            sb.append(currentChar);
            blocStart=i+1;
        }
    }
    if (sb != null) {
        if (length > blocStart) {
            // flush remaining characters
            sb.append(text, blocStart, length);
        }
        return sb.toString();
    }
    return text;
}`

For completeness, the `escapeRegex3` variant is:

`public static String escapeRegex3(String text) {
    int length = text.length();
    StringBuilder sb = new StringBuilder(length * 2); // prevent resizes
    int maxChar = CHAR_TO_ESCAPE.length;
    boolean escaped = false;
    for (int i = 0; i < length; i++) {
        char currentChar = text.charAt(i);
        if (currentChar < maxChar && CHAR_TO_ESCAPE[currentChar]) {
            sb.append('\\');
            escaped = true;
        }
        sb.append(currentChar);
    }
    return escaped ? sb.toString() : text;
}`